File: 剑指offer/28对称的二叉树.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func helper(p1, p2 *TreeNode) bool {
    if p1 == nil {
        return p2 == nil
    }
    if p2 == nil {
        return p1 == nil
    }
    if p1.Val != p2.Val {
        return false
    }
    return helper(p1.Left, p2.Right) && helper(p1.Right, p2.Left)
}

func isSymmetric(root *TreeNode) bool {
    if root == nil {
        return true
    }
    return helper(root.Left, root.Right)
}

File: 剑指offer/28对称的二叉树.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def helper(self, p1, p2) ->bool:
        if not p1:
            return not p2
        if not p2:
            return not p1
        if p1.val != p2.val:
            return False
        return self.helper(p1.left, p2.right) and self.helper(p1.right, p2.left)
    def isSymmetric(self, root: TreeNode) -> bool:
        if not root:
            return True
        return self.helper(root.left, root.right)